File: 1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero/1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero.cpp
class Solution {
public:
    int dfs(int source, vector<vector<int>>& adj ,vector<bool>& visited){
            int change = 0;
            visited[source]=true;
            for(auto dest:adj[source]){
                if(!visited[abs(dest)]){
                    change += dfs(abs(dest),adj,visited)+(dest>0);    
                }
            }
            
            return change;
    }
    int minReorder(int n, vector<vector<int>>& connections) {
        
        vector<vector<int>> adj(n,vector<int>());
        for (auto& c:connections){
            adj[c[0]].push_back(c[1]);
            adj[c[1]].push_back(-c[0]);
        }
        
        return dfs(0,adj,vector<bool>(n)={});
    }
};

File: 1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero/1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero.py
class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        gph = defaultdict(list)
        for u,v in connections:
            gph[u].append((v,True))
            gph[v].append((u,False))
        
        def solve(node,prev=-1):
            ans = 0
            for nex,dn in gph[node]:
                if nex!=prev:
                    ans += dn==True
                    ans+=solve(nex,node)
            return ans
        
        return solve(0)
    
    
    
    
    